vortex-layout/src/scan/executor/threads.rs

```diff
@@ -0,0 +1,96 @@
+use std::future::Future;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+
+use futures::channel::oneshot;
+use futures::future::BoxFuture;
+use futures::{FutureExt as _, TryFutureExt as _};
+use vortex_error::{vortex_err, VortexResult};
+
+use super::Executor;
+
+trait Task {
+    fn run(self: Box<Self>);
+}
+
+struct ExecutorTask<F, R> {
+    task: F,
+    result: oneshot::Sender<R>,
+}
+
+impl<F, R> Task for ExecutorTask<F, R>
+where
+    F: Future<Output = R> + Send,
+    R: Send,
+{
+    fn run(self: Box<Self>) {
+        let Self { task, result } = *self;
+        futures::executor::block_on(async move {
+            let output = task.await;
+            _ = result.send(output);
+        })
+    }
+}
+
+/// Multithreaded task executor, runs tasks on a dedicated thread pool.
+#[derive(Clone, Default)]
+pub struct ThreadsExecutor {
+    inner: Arc<Inner>,
+}
+
+impl ThreadsExecutor {
+    pub fn new(num_threads: NonZeroUsize) -> Self {
+        Self {
+            inner: Arc::new(Inner::new(num_threads)),
+        }
+    }
+}
+
+struct Inner {
```

Comment:
This structure is needed to achieve the behavior we want, where arbitrary futures can be spawned on this runtime. I've tried to structure it differently but I honestly think this is pretty nice, and we can definitely try and improve that later.